huaweicloud/services/smn/resource_huaweicloud_smn_message_detection.go

```diff
@@ -0,0 +1,268 @@
+			"endpoint": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `Specifies the endpoint address to be detected.`,
+			},
+			"extension": {
+				Type:        schema.TypeMap,
+				Optional:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: `Specifies the extended key/value for subscriptions over HTTP or HTTPS.`,
+			},
+			"result": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: `The message detection result.`,
+			},
+			"enable_force_new": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringInSlice([]string{"true", "false"}, false),
+				Description:  utils.SchemaDesc("", utils.SchemaDescInput{Internal: true}),
+			},
+		},
+	}
+}
+
+func resourceMessageDetectionCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	topicUrn := d.Get("topic_urn").(string)
+
+	client, err := cfg.NewServiceClient("smn", region)
+	if err != nil {
+		return diag.Errorf("error creating SMN client: %s", err)
+	}
+
+	// createMessageDetection: create SMN message detection
+	createMessageDetectionHttpUrl := "v2/{project_id}/notifications/topics/{topic_urn}/detection"
+	createMessageDetectionPath := client.Endpoint + createMessageDetectionHttpUrl
+	createMessageDetectionPath = strings.ReplaceAll(createMessageDetectionPath, "{project_id}", client.ProjectID)
+	createMessageDetectionPath = strings.ReplaceAll(createMessageDetectionPath, "{topic_urn}", topicUrn)
+
+	createMessageDetectionOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+	}
+
+	createMessageDetectionOpt.JSONBody = utils.RemoveNil(buildCreateMessageDetectionBodyParams(d))
+	createMessageDetectionResp, err := client.Request("POST", createMessageDetectionPath, &createMessageDetectionOpt)
+	if err != nil {
+		return diag.Errorf("error creating SMN message detection: %s", err)
+	}
+
+	createMessageDetectionRespBody, err := utils.FlattenResponse(createMessageDetectionResp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	id, err := jmespath.Search("task_id", createMessageDetectionRespBody)
+	if err != nil {
+		return diag.Errorf("error creating SMN message detection task: task ID is not found in API response")
+	}
+	d.SetId(id.(string))
+
+	// Check whether detection task has completed
+	err = checkDetectionTaskCompleted(ctx, client, topicUrn, d.Id(), d.Timeout(schema.TimeoutCreate))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return resourceMessageDetectionRead(ctx, d, meta)
+}
+
+func buildCreateMessageDetectionBodyParams(d *schema.ResourceData) map[string]interface{} {
+	bodyParams := map[string]interface{}{
+		"protocol":  d.Get("protocol"),
+		"endpoint":  d.Get("endpoint"),
+		"extension": buildDetectionExtensionBodyParams(d),
+	}
+	return bodyParams
+}
+
+func buildDetectionExtensionBodyParams(d *schema.ResourceData) map[string]interface{} {
```

Comment:
Maybe you can merge buildCreateMessageDetectionBodyParams and buildDetectionExtensionBodyParams into one function.